feat(agent): Fix GKE Autopilot installation

aroberts87 opened the pull request:
## What this PR does / why we need it:
There are several items related to the Agent chart that do not work when installing into a GKE Autopilot cluster. This change does the following to address them:

- Set a 500Mi ephemeral storage request on the agent container so that the containers are guaranteed enough space to compile the BPF probe
- Either take a priority class name, or optionally create a new priority class, to ensure the Agent pods get scheduled on the initial set of small nodes provided by Autopilot
- Enforce the use of the regular (non-slim) agent container in GKE Autopilot 

## Checklist

- [x] Title of the PR starts with type and scope, (e.g. feat(agent,node-analyzer,sysdig-deploy):)
- [x] Chart Version bumped for the respective charts
- [x] Variables are documented in the README.md (or README.tpl in some charts)
- [x] Check GithubAction checks (like lint) to avoid merge-check stoppers
- [x] All test files are added in the tests folder of their respective chart and have a "_test" suffix

aroberts87 committed:
Add ephemeral storage request in GKE Autopilot envs

In certain situations the agent would run into issues at startup if there was
not enough space to compile the eBPF probe. This change places an
ephemeral-storage request on the agent container when the gke.autopilot flag
has been set to address this issue.

saltycr3w reviewed (approved):
Looks good, and it's configurable, yay!

Did you you have to move the merge into the daemonset.yaml instead of inside the _helper for a particular reason?  Just curious.

dtaracks reviewed (approved):
So much better now - Thanks

aroberts87 commented:
> Looks good, and it's configurable, yay!
> 
> Did you you have to move the merge into the daemonset.yaml instead of inside the _helper for a particular reason? Just curious.

@dthomson-sysdig - Are you talking about for the kmodule container? I left it in the daemonset instead of moving to the _helper file because those values were just being taken directly from the `values.yaml` and don't ever need to comply with the resource profiles, so I figured it would reduce the complexity to just handle the splicing locally. Does that make sense?

aroberts87 committed:
Ensure we're placing resourceProfile limits again.

The initial commit only applied the requests and erroneously omitted
the limits block.

aroberts87 committed:
Add additional unit tests to check resource limits are set correctly

aroberts87 committed:
Fix GKE Autopilot installation

* If gke.autopilot is enabled we will now optionally create a PriorityClass
  object to ensure that the Agent Pods can be scheduled on the small initial
  nodes.
* If gke.autopilot is enabled the daemonset resources will have their requests
  adjusted to be low enough to be scheduled on the inital set of nodes
* Actively fail chart rendering if slim.enabled and gke.autopilot are enabled
  as the slim containers will not run in GKE Autopilot environments

mavimo committed:
fix(agent,sysdig): do not mount /lib/modules and /usr when GKE autopilot is enabled

mavimo commented on charts/agent/templates/daemonset.yaml:
```diff
@@ -1,4 +1,7 @@
 {{- if .Values.daemonset.deploy }}
+{{- if and .Values.gke.autopilot .Values.slim.enabled }}
```
use `{{- include "agent.gke.autopilot" . }}` to check for autopilot using the helper.

mavimo commented on charts/agent/templates/daemonset.yaml:
```diff
@@ -29,8 +32,8 @@ spec:
       serviceAccountName: {{ template "agent.serviceAccountName" .}}
-{{- if .Values.priorityClassName }}
-      priorityClassName: "{{ .Values.priorityClassName }}"
+{{- if or .Values.priorityClassName .Values.gke.autopilot }}
```
use `{{- include "agent.gke.autopilot" . }}` to check for autopilot using the helper.

mavimo commented on charts/agent/templates/gkeautopilotpriorityclass.yaml:
```diff
@@ -0,0 +1,16 @@
+{{- if and .Values.gke.autopilot (not .Values.slim.enabled) }}
```
use `{{- include "agent.gke.autopilot" . }}` to check for autopilot using the helper.

aroberts87 committed:
Remove unit tests for failed renderings that no longer happen

aroberts87 committed:
Convert fail to warn message for missing PriorityClass creation or name specification

aroberts87 committed:
Improve GKE Autopilot detection

mavimo committed:
chore(agent): use the proper way to quote variables in template

mavimo committed:
fix(agent): mount only /var/run/containerd/containerd.sock when agent work on slim mode on GKE Autopilot environment

mavimo committed:
feat(agent): mount only the folder required when running GKE Autopilot & Slim agent

mavimo committed:
feat(agent): remove unrequired mount volumes

mavimo committed:
feat(agent): not mount /etc/modprobe.d when running on GKE Autopilot

mavimo committed:
feat(agent): do not mount the /sys/kernel/debug in GKE Autopilot

mavimo committed:
feat(agent): /etc/os-release is always available in COS (GKE Autopilot use COS)

mavimo committed:
feat(agent): privileged containers make the capabilities drop irrelevant

mavimo committed:
feat(agent): mount only the folder required when running GKE Autopilot & Slim agent

feat(agent): remove a file added by mistake.

feat(agent): Remove test file

remove files adde d by mistake

aroberts87 commented on post-change-3.yaml:
```diff
@@ -0,0 +1,469 @@
+---
```
Was this file intended to be committed?

aroberts87 commented on pre-change-3.yaml:
```diff
@@ -0,0 +1,486 @@
+---
```
Same as above, was this file intended to be committed?

aroberts87 commented on charts/agent/templates/gkeautopilotpriorityclass.yaml:
```diff
@@ -0,0 +1,13 @@
+{{- if and (include "agent.gke.autopilot" .) (not .Values.slim.enabled) .Values.gke.createPriorityClass }}
```
Should we remove the constraint around slim not being enabled now?

aroberts87 commented on charts/agent/templates/deployment.yaml:
```diff
@@ -105,28 +102,13 @@ spec:
+          {{- /* Always requested */}}
             - mountPath: /host/dev
               name: dev-vol
+              readOnly: false
```
What is the reason for this change?

mavimo replied:
Is the default value, I added it supposing that make it more clear, but let's remove it 😅

aroberts87 commented on charts/agent/templates/daemonset.yaml:
```diff
@@ -72,32 +72,29 @@ spec:
           image: {{ template "agent.image.kmodule" . }}
           imagePullPolicy: {{ .Values.image.pullPolicy }}
           securityContext:
-            capabilities:
-              drop:
-                - ALL
```
What is the purpose of this change?

mavimo replied:
We're specifying this need to run as a privileged container, so the capabilities drop won't have any effect.

aroberts87 commented on charts/agent/templates/daemonset.yaml:
```diff
@@ -112,45 +109,51 @@ spec:
               value: {{ .Values.proxy.noProxy | default .Values.global.proxy.noProxy }}
           {{- end }}
           volumeMounts:
+          {{- /* Slim = true, Autopilot = false */}}
+          {{- if not (include "agent.gke.autopilot" .) }}
```
We're not checking the value of `slim.enabled` here, so won't this be picked in both cases?

mavimo replied:
Is by default since the init container is available only in slim mode (see [line 77](https://github.com/sysdiglabs/charts/pull/975/files#diff-18888d7372d81c06ea1ec38c2606ae16bf04d0248bdb8ad475cbc5cc22bbc364R77))

aroberts87 commented on charts/agent/templates/daemonset.yaml:
```diff
@@ -112,45 +109,51 @@ spec:
-{{ include "agent.resources" . | indent 12 }}
+            {{- include "agent.resources" . | nindent 12 }}
           securityContext:
-            capabilities:
-              drop:
-                - ALL
```
What is the reason for this change?

mavimo replied:
We're specifying this need to run as a privileged container, so the capabilities drop won't have any effect.

mavimo committed:
fix(agent,sysdig-deploy): bump charts
